src/abi_arm.cpp

```diff
@@ -0,0 +1,315 @@
+        return NULL;
+    Type *lltype;
+    // Check size first since it's cheaper.
+    switch (dt->size) {
+    case 2:
+        lltype = T_float16;
+        break;
+    case 4:
+        lltype = T_float32;
+        break;
+    case 8:
+        lltype = T_float64;
+        break;
+    default:
+        return NULL;
+    }
+    return jl_is_floattype((jl_value_t*)dt) ? lltype : NULL;
+}
+
+static size_t isContainerizedVector(jl_datatype_t *dt)
+{
+    // "Containerized Vector" in the ARM AAPCS refers to fundamental data
+    // types (machine types), it's the class of machine types comprising
+    // the 64-bit and 128-bit NEON machine types.
+
+    // TODO
+
+    return 0;
+}
+
+static size_t isLegalHA(jl_datatype_t *dt, Type *&base);
+
+// Check whether a type contained by a candidate homogeneous aggregate is valid.
+// Returns the corresponding LLVM type.
+static Type *isLegalHAType(jl_datatype_t *dt)
+{
+    // single- or double-precision floating-point type
+    Type* fp = get_llvm_fptype(dt);
+    if (fp)
+        return fp;
+
+    // 64- or 128-bit containerized vectors
+    if (isContainerizedVector(dt)) {
+        // TODO
+    }
+
+    // recursive application (composite types can contain other composites)
+    Type *base = NULL;
+    if (isLegalHA(dt, base))
+        return base;
+
+    return NULL;
+}
+
+// Check whether a type is a legal homogeneous aggregate. Returns the number of
+// members.
+static size_t isLegalHA(jl_datatype_t *dt, Type *&base) {
+    // Homogeneous aggregates are only used for VFP registers,
+    // so use that definition of legality (section 6.1.2.1)
+
+    if (jl_is_structtype(dt)) {
+        // ... with one to four Elements.
+        size_t members = jl_datatype_nfields(dt);
+        if (members < 1 || members > 4)
+            return 0;
+
+        base = NULL;
+        for (size_t i = 0; i < members; ++i) {
+            Type *T = isLegalHAType((jl_datatype_t*)jl_field_type(dt,i));
+            if (!T)
+                return 0;
+
+            if (!base)
+                base = T;
+            else if (base != T)
+                return 0;
+        }
+        return members;
+    }
+
+    return 0;
+}
+
+// Determine if an argument can be passed through a coprocessor register.
+//
+// All the out parameters should be default to `false`.
+static void classify_cprc(jl_datatype_t *dt, bool *vfp)
+{
+    // Based on section 6.1 of the Procedure Call Standard
+
+    // VFP: 6.1.2.1
+    // - A half-precision floating-point type.
+    // - A single-precision floating-point type.
+    // - A double-precision floating-point type.
+    if (get_llvm_fptype(dt)) {
+        *vfp = true;
+        return;
+    }
+
+    // - A 64-bit or 128-bit containerized vector type.
+    size_t vec_size = isContainerizedVector(dt);
+    if (vec_size == 8 || vec_size == 16) {
+        *vfp = true;
+        return;
+    }
+
+    // - A Homogeneous Aggregate
+    Type *base = NULL;
+    if (isLegalHA(dt, base)) {
```

Comment:
The AAPCS doc is much harder to read indeed, some sections just seem out of order... I guess they've build on that experience when drafting the AArch doc.

Anyway, I think some aspects of the more complex legality definition are needed though, for example handling of nested composites with the same type (this happens in my extended test suite containing `struct { complex{T}, complex{T} }`). I also thought that it might be interesting to keep the definition a bit more general, in order to share it between ABI's later on (mimicking Clang here).
That said, I don't have particulary strong feelings, so if you prefer a simpler, stand-alone ABI definition I'm willing to simplify it.